### PR TITLE
Feature/refactoring internship offer

### PR DESCRIPTION
**Pull request by Moussa348**
1.refactoring of internshipOffer, dto,mapper,service,controller
2.refactoring monitor repository, service
3.replacing ResponseStatusException with custom Exception
4.adding ResponseEntity on all controller

### CODE REVIEW

File: backend/src/main/java/com/team4/backend/exception/DoNotExistException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.team4.backend.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+public class DoNotExistException extends Exception {
```
Comment (0xD34DC0DE) [inline backend/src/main/java/com/team4/backend/exception/DoNotExistException.java:6]: suggestion: more appropriate name -> UserNotFoundException

File: backend/src/main/java/com/team4/backend/service/MonitorService.java [hunk 2]
```diff
@@ -26,10 +24,10 @@ public MonitorService(MonitorRepository monitorRepository, UserService userServi
...
     }
 
-    public Mono<Monitor> findMonitorByEmail(String email) {
-        return monitorRepository.findByEmail(email)
-                .filter(Objects::nonNull)
-                .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND, "Can't find monitor with this email.")));
+    public Mono<Boolean> existsByEmailAndIsEnabledTrue(String email) {
+        return monitorRepository.existsByEmailAndIsEnabledTrue(email)
+                .filter(exist -> exist)
+                .switchIfEmpty(Mono.error(DoNotExistException::new));
```
Comment (0xD34DC0DE) [inline backend/src/main/java/com/team4/backend/service/MonitorService.java:30]: Why filter return an exception when False is the intended behavior ?

File: backend/src/main/java/com/team4/backend/util/JwtUtil.java [hunk 3]
```diff
@@ -35,12 +36,19 @@ public void init() {
...
         final Date expirationDate = new Date(createdDate.getTime() + expirationTimeLong * 1000);
 
-        claims.put("role", user.getRole());
+        claims.put("role", user.getRole().toString());
+        claims.put("firstName", user.getFirstName());
+        claims.put("lastName", user.getLastName());
+        claims.put("phoneNumber", user.getPhoneNumber());
+
+        if (user.getRole().equals(Role.MONITOR))
+            claims.put("companyName", ((Monitor) user).getCompanyName());
```
Comment (0xD34DC0DE) [inline backend/src/main/java/com/team4/backend/util/JwtUtil.java:50]: generate token should always return the same thing regardless of Role, I think a better solution is needed.
Call me on discord when you read this.

File: backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java [hunk 4]
```diff
@@ -43,7 +42,25 @@ void addAnInternshipOffer() {
                 .bodyValue(internshipOfferDTO)
                 .exchange()
                 //ASSERT
-                .expectStatus().isOk()
+                .expectStatus().isCreated()
+                .expectBody(InternshipOfferDto.class);
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java:46]: Previous PR comment will change this

File: backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java [hunk 5]
```diff
@@ -43,7 +42,25 @@ void addAnInternshipOffer() {
...
+
+        //ACT
+        webTestClient
+                .post()
+                .uri("/internshipOffer/addAnInternshipOffer")
+                .bodyValue(internshipOfferDTO)
+                .exchange()
+                //ASSERT
+                .expectStatus().isNotFound()
                 .expectBody(InternshipOfferDto.class);
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java:64]: same comment

**Review by 0xD34DC0DE — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/team4/backend/service/MonitorService.java (hunk 2 above)
Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/service/MonitorService.java:30]: because it will go through without throwing an exception

Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/service/MonitorService.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/e4d3c34bf0758d1781144b35a26d663b16a4fc9c

File: backend/src/main/java/com/team4/backend/exception/DoNotExistException.java (hunk 1 above)
Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/exception/DoNotExistException.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/b4dcf8d4f6d6dfebd5ae5eb83a434ae18094769f

**Review by wjeam — APPROVED** [review]

File: backend/src/main/java/com/team4/backend/controller/InternshipOfferController.java [hunk 6]
```diff
@@ -21,7 +24,9 @@ public InternshipOfferController(InternshipOfferService internshipOfferService)
 
     @PostMapping("/addAnInternshipOffer")
     @PreAuthorize("hasAnyAuthority('INTERNSHIP_MANAGER','MONITOR')")
-    public Mono<InternshipOfferDto> addAnInternshipOffer(@RequestBody InternshipOfferDto internshipOfferDTO){
-        return internshipOfferService.addAnInternshipOffer(internshipOfferDTO);
+    public Mono<ResponseEntity<InternshipOfferDto>> addAnInternshipOffer(@RequestBody InternshipOfferDto internshipOfferDTO) {
+        return internshipOfferService.addAnInternshipOffer(internshipOfferDTO)
+                .flatMap(internshipOffer -> Mono.just(ResponseEntity.status(HttpStatus.CREATED).body(internshipOffer)))
```
[the remark this answers is not in the document]
Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/controller/InternshipOfferController.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/30f4cdfbc8b32fb97f715d5e7556938d02a01b54

File: backend/src/main/java/com/team4/backend/controller/StudentController.java [hunk 7]
```diff
@@ -30,7 +30,7 @@ public StudentController(StudentService studentService) {
     public Mono<ResponseEntity<String>> register(@RequestBody StudentDto studentDto) {
         return studentService.registerStudent(StudentMapper.toEntity(studentDto))
```
[the remark this answers is not in the document]
Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/controller/StudentController.java:28]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/d2ed1c85f6be22103575d99fd05bb075e89feae9

File: backend/src/main/java/com/team4/backend/controller/UserController.java [hunk 8]
```diff
@@ -16,13 +19,16 @@ public UserController(UserService userService){
     }
 
     @PostMapping("/login")
-    public Mono<String> login(@RequestBody AuthRequestDto authRequestDto){
-        return userService.login(authRequestDto);
+    public Mono<ResponseEntity<String>> login(@RequestBody AuthRequestDto authRequestDto){
```
[the remark this answers is not in the document]
Comment (Moussa348) [reply backend/src/main/java/com/team4/backend/controller/UserController.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/d2ed1c85f6be22103575d99fd05bb075e89feae9

File: backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java (hunk 4 above)
Comment (Moussa348) [reply backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/d2ed1c85f6be22103575d99fd05bb075e89feae9

File: backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java (hunk 5 above)
Comment (Moussa348) [reply backend/src/test/java/com/team4/backend/controller/InternshipOfferControllerTest.java]: https://github.com/0xD34DC0DE/ProjetIntegreTeam4/pull/18/commits/d2ed1c85f6be22103575d99fd05bb075e89feae9

**Comment by 0xD34DC0DE** [comment]
I'm going to merge it so I can continue work but don't delete the branch yet there are still things to change

**Review by 0xD34DC0DE — APPROVED** [review]

**Pull request merged**